### PR TITLE
[SYCLomatic] Trigger intercept-build tool from DPCT

### PR DESCRIPTION
**Pull request by daiyaan-ahmed6**
[Feature]: Added option for SYCLomatic customized intercept-build tool to be called as a DPCT command line option.
E2E Test added for option: https://github.com/oneapi-src/SYCLomatic-test/pull/464
Signed-off-by: Daiyaan Ahmed <daiyaan.ahmed@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/DPCT.cpp [hunk 1]
```diff
@@ -614,6 +615,31 @@ int runDPCT(int argc, const char **argv) {
...
+    llvm::sys::path::append(pathToInterceptBuildBinary, "bin",
+                            "intercept-build");
+    if (!llvm::sys::fs::exists(pathToInterceptBuildBinary)) {
+      DpctLog() << "Error: intercept-build tool not found"
+                << "\n";
+      ShowStatus(MigrationErrorInvalidInstallPath);
+      dpctExit(MigrationErrorInvalidInstallPath);
+    }
+    std::string interceptBuildSystemCall =
+        (pathToInterceptBuildBinary + " " + InterceptBuildCommand).str();
```
Comment (tomflinda) [inline clang/lib/DPCT/DPCT.cpp]: From the implementation, it only supports the fixed format of "dpct --intercept-build make".
While for the options of intercept-build tool, like:
```
optional arguments:
  -h, --help            show this help message and exit
  --verbose, -v         Enable verbose output from 'intercept-build'. A second, third and fourth flags increases verbosity. (default: 0)
  --cdb <file>          The JSON compilation database. (default: compile_commands.json)
  --parse-build-log <file>
                        Specifies the file path of the build log. (default: None)
  --work-directory <path>
                        Specifies the working directory of the command that generates the build log specified by option --parse-build-log. (default: the directory of build log file specified by option
                        --parse-build-log)
  --linker-entry        Generate linker entry in compilation database if '--linker-entry' option is present. (default: 1)
  --no-linker-entry     Do not generate linker entry in compilation database if '--no-linker-entry' option is present. (default: 0)

advanced options:
  --append              Extend existing compilation database with new entries. Duplicate entries are detected and not present in the final output. The output is not continuously updated, it's done when
                        the build command finished. (default: False)
```
and options for "make" tool:
```
  -b, -m                      Ignored for compatibility.
  -B, --always-make           Unconditionally make all targets.
  -C DIRECTORY, --directory=DIRECTORY
                              Change to DIRECTORY before doing anything.
  -d                          Print lots of debugging information.
  --debug[=FLAGS]             Print various types of debugging information.
  -e, --environment-overrides
                              Environment variables override makefiles.
...
```
the command "dpct --intercept-build" should also support.
For example: "dpct --intercept-build -vvv make", "dpct --intercept-build make -B", "dpct --intercept-build -vvv make -B" should be supported.

File: clang/lib/DPCT/DPCT.cpp [hunk 2]
```diff
@@ -614,6 +615,31 @@ int runDPCT(int argc, const char **argv) {
     dpctExit(MigrationSucceeded);
   }
 
+#ifndef _WIN32
+  if (!InterceptBuildCommand.empty()) {
+    SmallString<512> pathToInterceptBuildBinary(DpctInstallPath);
```
Comment (ShengchenJ) [inline clang/lib/DPCT/DPCT.cpp]: Algin the code style。
```suggestion
    SmallString<512> PathToInterceptBuildBinary(DpctInstallPath);
```

File: clang/lib/DPCT/DPCT.cpp [hunk 3]
```diff
@@ -614,6 +615,31 @@ int runDPCT(int argc, const char **argv) {
...
+    llvm::sys::path::append(pathToInterceptBuildBinary, "bin",
+                            "intercept-build");
+    if (!llvm::sys::fs::exists(pathToInterceptBuildBinary)) {
+      DpctLog() << "Error: intercept-build tool not found"
+                << "\n";
+      ShowStatus(MigrationErrorInvalidInstallPath);
+      dpctExit(MigrationErrorInvalidInstallPath);
+    }
+    std::string interceptBuildSystemCall =
```
Comment (ShengchenJ) [inline clang/lib/DPCT/DPCT.cpp]:
```suggestion
    std::string InterceptBuildSystemCall =
```

**Review by ShengchenJ — CHANGES_REQUESTED** [review]

File: clang/lib/DPCT/DPCT.cpp (hunk 1 above)
Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp]: Fixed by taking direct command line arguments from argv.

Comment (tomflinda) [reply clang/lib/DPCT/DPCT.cpp]: @daiyaan-ahmed6 “dpct --intercept-build -vvv make” has been supported while options for make still not supported. i.e.: "dpct --intercept-build make -B" still not supported。

Also the note msg should not be emitted during running "dpct --intercept-build -v /usr/bin/make".
```
Chenwei@chenweis-Precision-7920-Tower:~/NVIDIA_CUDA-11.4_Samples/0_Simple/vectorAdd$ dpct --intercept-build -v /usr/bin/make
NOTE: /usr/bin/make is treated as a CUDA file by default. Use the --extra-arg=-xc++ option to treat /usr/bin/make as a C++ file if needed.
NOTE: Could not auto-detect compilation database for file 'make' in '/usr/bin' or any parent directory.
>>> GCC Version is greater or equal to 4.7.0 <<<
make: Nothing to be done for 'all'.

Chenwei@chenweis-Precision-7920-Tower:~/NVIDIA_CUDA-11.4_Samples/0_Simple/vectorAdd$ dpct --intercept-build -v /usr/bin/make -B
Unknown command line argument '-B'. Did you mean '-h'?
dpct exited with code: -16 (Option parsing error, run 'dpct --help' to see supported options and values)

Chenwei@chenweis-Precision-7920-Tower:~/NVIDIA_CUDA-11.4_Samples/0_Simple/vectorAdd$
```

Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp]: ~~"dpct --intercept-build -v make -B" is not being supported as LLVM command line interprets `-` as direct argument to dpct.~~

~~Reference: https://llvm.org/docs/CommandLine.html#specifying-positional-options-with-hyphens~~

~~However, "dpct --intercept-build -v make -- -B" works. Do you think implementing it this way is good? I can specify in the description for intercept-build command line option to use `--` to pass positional arguments to build command.~~

Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp]: Removed the "Note" warning on calling intercept-build

Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp]: > ~"dpct --intercept-build -v make -B" is not being supported as LLVM command line interprets `-` as direct argument to dpct.~
> 
> ~Reference: https://llvm.org/docs/CommandLine.html#specifying-positional-options-with-hyphens~
> 
> ~However, "dpct --intercept-build -v make -- -B" works. Do you think implementing it this way is good? I can specify in the description for intercept-build command line option to use `--` to pass positional arguments to build command.~

Was able to bypass this with llvm::cl::Sink command line option.

File: clang/lib/DPCT/DPCT.cpp (hunk 2 above)
Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp]: Done.

File: clang/lib/DPCT/DPCT.cpp (hunk 3 above)
Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp]: Done.

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by ShengchenJ — APPROVED** [review]

File: clang/lib/DPCT/DPCT.cpp [hunk 4]
```diff
@@ -614,6 +615,34 @@ int runDPCT(int argc, const char **argv) {
...
+    std::string InterceptBuildSystemCall(PathToInterceptBuildBinary.str());
+    for (int argumentIndex = 2; argumentIndex < argc; argumentIndex++) {
+      InterceptBuildSystemCall.append(" ");
+      InterceptBuildSystemCall.append(std::string(argv[argumentIndex]));
+    }
+    int processExitCode = system(InterceptBuildSystemCall.c_str());
+    if (processExitCode) {
+      DpctLog() << "Error: intercept-build process call not successful"
+                << "\n";
+      ShowStatus(MigrationError);
```
Comment (zhimingwang36) [inline clang/lib/DPCT/DPCT.cpp]: we may define a new error code for intercept-build.

File: clang/lib/Tooling/CommonOptionsParser.cpp [hunk 5]
```diff
@@ -176,6 +176,11 @@ OPT_TYPE OPT_VAR(OPTION_NAME, __VA_ARGS__);
 
   SourcePathList = SourcePaths;
 #ifdef SYCLomatic_CUSTOMIZATION
+#ifndef _WIN32
+  if (std::string(argv[1]) == "--intercept-build") {
```
Comment (zhimingwang36) [inline clang/lib/Tooling/CommonOptionsParser.cpp]: how about make "-intercept-build" and "intercept-build" also works.

File: llvm/lib/Support/CommandLine.cpp [hunk 6]
```diff
@@ -1537,6 +1537,23 @@ bool CommandLineParser::ParseCommandLineOptions(int argc,
   GlobalParser->ActiveSubCommand = ChosenSubCommand;
 
   assert(ChosenSubCommand);
+
+#ifdef SYCLomatic_CUSTOMIZATION
+#ifndef _WIN32
+  if (std::string(argv[FirstArg]) == "--intercept-build") {
```
Comment (zhimingwang36) [inline llvm/lib/Support/CommandLine.cpp]: how about make "-intercept-build" or "intercept-build" also works.

**Review by zhimingwang36 — COMMENTED** [review]
Please add a lit test, to show the new option is working.

File: clang/lib/DPCT/DPCT.cpp (hunk 4 above)
Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp:638]: Could you suggest an error code name for this please?

Do you think `ProcessError` / `SubcommandError` / `SubCommandExecutionFailure` is suitable?

Comment (daiyaan-ahmed6) [reply clang/lib/DPCT/DPCT.cpp]: Defined new Error code as `InterceptBuildSuccess` / `InterceptBuildError`. Done.

File: clang/lib/Tooling/CommonOptionsParser.cpp (hunk 5 above)
Comment (daiyaan-ahmed6) [reply clang/lib/Tooling/CommonOptionsParser.cpp]: Modified and tested to make `-intercept-build` and `intercept-build` to also work. Done.

File: llvm/lib/Support/CommandLine.cpp (hunk 6 above)
Comment (daiyaan-ahmed6) [reply llvm/lib/Support/CommandLine.cpp]: Modified and tested to make `-intercept-build` and `intercept-build` to also work. Done.

**Comment by daiyaan-ahmed6** [comment]
> Please add a lit test, to show the new option is working.

LIT test for intercept-build does not work. Instead added E2E test in SYCLomatic-test repo to ensure `intercept-build` works as required.
https://github.com/oneapi-src/SYCLomatic-test/pull/464

File: clang/tools/scan-build-py/CMakeLists.txt [hunk 7]
```diff
@@ -167,11 +167,11 @@ if (SYCLomatic_CUSTOMIZATION)
     )
 
     install(TARGETS intercept-stub
-            DESTINATION lib/libear
+            DESTINATION ${CMAKE_BINARY_DIR}/lib/libear
```
Comment (tomflinda) [inline clang/tools/scan-build-py/CMakeLists.txt]: This line and line174 changes original logic, which make the intercept-stub only installed in the build directory, not the installation directory, which shoud be the root cause of CI test failure of cmp-cmds-linker-entry-src-files.

The change should be add a new operation to copy intercept-stub into the lib/libear of the build directory, and the original copy should be kept.

Comment (daiyaan-ahmed6) [reply clang/tools/scan-build-py/CMakeLists.txt]: Done.

**Review by tomflinda — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**